lib/cfn-nag/util/iam_permutations.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Find all pemutations of an IAM string with wildcards approprate to the location
+
+def iam_permutations(string, element, prefix: '', suffix: '')
```

Comment:
i'd say everything looks good in this PR except for this guy.  most important review comment is that this is very error prone code that lends itself wonderfully to automated unit tests..... but i don't see any, so that's my main gripe.

i have some advice on the interface and the implementation.  

for the interface.... it's not clear to me why the "action", prefix or suffix are needed.  i'd expect to pass in a string like PassRole and get the collection of wildcards back (and then up to me to add iam: if I so choose.  On the other hand, wildcards behave differently by *service*... but i wouldn't sweat that just yet.

I honestly didn't try to decipher the implementation in great depth since it didn't have the tests to read first.  that said, i don't know for sure, but instead of saying "i bet" - let's say - "it wouldn't surprise me if...." a recursive solution would be about half the code and read more concisely.

Reply:
I converted it to recursive and made it a generic `wildcard_patterns`.  adding the `iam:` at the front is taken care of now with a map.

To give wildcard_patterns flexibility I have an optional `patterns` parameter with the default being to include wildcards before, after and both before and after.

Reply:
Recursive is definitely much clearer / less code.